File: sidecar/transcribe.py
"""memolo 音声文字起こし sidecar.

stdin: 改行終端の JSON 1行
  {"wav_path": "/abs/path.wav", "language": "ja", "model_size": "medium"}
  - model_size は env MEMOLO_WHISPER_MODEL があればそれで上書き

stdout: JSON-Lines (1行=1イベント, ensure_ascii=False, flush=True)
  {"type": "loaded",  "model": "medium", "device": "cpu"}
  {"type": "info",    "duration": 18.4, "language": "ja", "language_probability": 0.99}
  {"type": "segment", "id": 0, "start": 0.0, "end": 2.4, "text": "..."}
  ...
  {"type": "done",    "total_segments": 12, "full_text": "..."}
  {"type": "error",   "message": "..."}   # 例外時 (exit 1)

stderr: faster-whisper 自身のログを含むデバッグ出力
"""

from __future__ import annotations

import json
import os
import sys
import traceback


def emit(event: dict) -> None:
    print(json.dumps(event, ensure_ascii=False), flush=True)


def main() -> int:
    raw = sys.stdin.readline()
    if not raw.strip():
        emit({"type": "error", "message": "no config on stdin"})
        return 1

    try:
        cfg = json.loads(raw)
    except json.JSONDecodeError as e:
        emit({"type": "error", "message": f"stdin JSON parse: {e}"})
        return 1

    wav_path = cfg.get("wav_path")
    if not wav_path:
        emit({"type": "error", "message": "wav_path is required"})
        return 1

    model_size = os.environ.get("MEMOLO_WHISPER_MODEL") or cfg.get("model_size", "medium")
    language = cfg.get("language", "ja")

    try:
        from faster_whisper import WhisperModel
    except ImportError as e:
        emit({"type": "error", "message": f"faster-whisper not installed: {e}"})
        return 1

    try:
        model = WhisperModel(model_size, device="cpu", compute_type="int8")
        emit({"type": "loaded", "model": model_size, "device": "cpu"})

        segments_iter, info = model.transcribe(
            wav_path,
            language=language,
            vad_filter=True,
        )
        emit({
            "type": "info",
            "duration": float(info.duration),
            "language": info.language,
            "language_probability": float(info.language_probability),
        })

        texts: list[str] = []
        for i, seg in enumerate(segments_iter):
            texts.append(seg.text)
            emit({
                "type": "segment",
                "id": i,
                "start": float(seg.start),
                "end": float(seg.end),
                "text": seg.text,
            })

        emit({"type": "done", "total_segments": len(texts), "full_text": "".join(texts)})
        return 0
    except Exception as e:
        traceback.print_exc(file=sys.stderr)
        emit({"type": "error", "message": f"{type(e).__name__}: {e}"})
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: src-tauri/src/transcribe.rs
use std::path::PathBuf;
use std::process::Stdio;

use serde::{Deserialize, Serialize};
use tauri::ipc::Channel;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::Command;

#[derive(Serialize, Deserialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum TranscribeEvent {
    Loaded {
        model: String,
        device: String,
    },
    Info {
        duration: f64,
        language: String,
        language_probability: f64,
    },
    Segment {
        id: u32,
        start: f64,
        end: f64,
        text: String,
    },
    Done {
        total_segments: u32,
        full_text: String,
    },
    Error {
        message: String,
    },
}

fn resolve_sidecar() -> Result<(PathBuf, PathBuf), String> {
    // dev 時の current_dir は src-tauri/。その親 (= memolo/) 配下の sidecar/ を見る。
    let cwd = std::env::current_dir().map_err(|e| format!("cwd: {e}"))?;
    let project_root = cwd
        .parent()
        .ok_or_else(|| "cwd has no parent".to_string())?
        .to_path_buf();
    let python = project_root.join("sidecar/.venv/bin/python");
    let script = project_root.join("sidecar/transcribe.py");
    Ok((python, script))
}

#[tauri::command]
pub async fn transcribe(
    wav_path: String,
    on_event: Channel<TranscribeEvent>,
) -> Result<(), String> {
    let (python, script) = resolve_sidecar()?;

    if !python.exists() {
        return Err(format!(
            "sidecar venv が見つかりません: {}\n\
             プロジェクト直下で `cd sidecar && uv venv --python 3.11 && uv sync` を実行してください。",
            python.display()
        ));
    }
    if !script.exists() {
        return Err(format!("sidecar スクリプトが見つかりません: {}", script.display()));
    }

    let mut child = Command::new(&python)
        .arg(&script)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .kill_on_drop(true)
        .spawn()
        .map_err(|e| format!("spawn sidecar: {e}"))?;

    let cfg = serde_json::json!({
        "wav_path": wav_path,
        "language": "ja",
    });
    {
        let mut stdin = child
            .stdin
            .take()
            .ok_or_else(|| "no stdin".to_string())?;
        stdin
            .write_all(format!("{}\n", cfg).as_bytes())
            .await
            .map_err(|e| format!("write stdin: {e}"))?;
        stdin
            .flush()
            .await
            .map_err(|e| format!("flush stdin: {e}"))?;
    } // drop stdin → EOF を sidecar に通知

    let stdout = child
        .stdout
        .take()
        .ok_or_else(|| "no stdout".to_string())?;
    let mut lines = BufReader::new(stdout).lines();
    while let Some(line) = lines
        .next_line()
        .await
        .map_err(|e| format!("read stdout: {e}"))?
    {
        if line.trim().is_empty() {
            continue;
        }
        match serde_json::from_str::<TranscribeEvent>(&line) {
            Ok(ev) => {
                if let Err(e) = on_event.send(ev) {
                    eprintln!("channel send: {e}");
                }
            }
            Err(e) => {
                eprintln!("sidecar JSON parse error: {e} line={line}");
            }
        }
    }

    let status = child.wait().await.map_err(|e| format!("wait sidecar: {e}"))?;
    if !status.success() {
        return Err(format!("sidecar exited with {status}"));
    }
    Ok(())
}
